### PR TITLE
bump uSockets to 0.8.1 with eventloop option

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **usockets/0.8.1**

uSockets/0.8.0 begins to support boost.asio as event loop library. 
So I add 'eventloop' option instead of 'with_libuv'.

I'm afraid package options is breaked compatibility.
But I suggest to delete 'with_libuv'. 
Because libuv is one of the event loop implementations in usockets. 
(others are epolll, kqueue, gcd, asio.)

I add compiler version check too.
usockets uses C++14 with 'asio' options.
usockets uses C++17 with 'openssl' options. (It may change future release)

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by toge** [comment]
CI failed seems to be caused by Visual Studio license expiration.
What can I do for it?

```
usockets/0.8.1: Calling build()

----Running------
> set "VSCMD_START_DIR=%CD%" && call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC/Auxiliary/Build/vcvarsall.bat" amd64 && devenv "uSockets.vcxproj" /upgrade && msbuild "uSockets.vcxproj" /p:Configuration="Release" /p:UseEnv=true /p:Platform="x64" /m:16 /p:PlatformToolset="v141" /verbosity:minimal /p:ForceImportBeforeCppTargets="C:\J\w\BuildSingleReference@5\.conan\data\usockets\0.8.1\_\_\build\9514e1b97a672267232de29ef5b23bb442333a1d\source_subfolder\conan_build.props"
-----------------
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.9.19
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

Microsoft Visual Studio 2017 Version 15.0.28307.1000.
Copyright (C) Microsoft Corp. All rights reserved.

The license for Visual Studio has expired.

The evaluation period for this product has ended.usockets/0.8.1: 
```

File: recipes/usockets/all/conanfile.py [hunk 1]
```diff
@@ -14,8 +14,10 @@ class UsocketsConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
     options = {"fPIC": [True, False],
                "with_ssl": [False, "openssl", "wolfssl"],
-               "with_libuv": [True, False]}
-    default_options = {"fPIC": True, "with_ssl": False, "with_libuv": True}
+               "eventloop": ["default", "libuv", "gcd", "asio"]}
```
Comment (madebr) [inline recipes/usockets/all/conanfile.py:17]: Don't add a `default` option value.
Use one in `default_options`.

It is always possible to override the default in `config_options`:
e.g.
```python
default_config_options = {"eventloop": "asio"}
def config_options(self):
    if self.settings.os == "Windows":
        self.options.eventloop = "libuv"
```

File: recipes/usockets/all/conanfile.py [hunk 2]
```diff
@@ -26,17 +28,77 @@ def config_options(self):
...
-            self.requires("libuv/1.41.0")
+            self.requires("wolfssl/4.8.1")
+
+        if self.options.eventloop == "default" and self.settings.os == "Windows":
+            self.requires("libuv/1.41.1")
+        elif self.options.eventloop == "libuv":
+            self.requires("libuv/1.41.1")
+        elif self.options.eventloop == "gcd":
+            self.requires("libdispatch/5.3.2")
+        elif self.options.eventloop == "asio":
```
Comment (madebr) [inline recipes/usockets/all/conanfile.py:100]: Perhaps this option should be named `boost` instead?
cci also has a `asio` recipe.

File: recipes/usockets/all/conanfile.py [hunk 3]
```diff
@@ -26,17 +28,77 @@ def config_options(self):
...
+
+        return version
+
+    def validate(self):
+        if self.options.eventloop == "gcd":
+            if self.settings.os != "Linux" or self.settings.compiler != "clang" :
+                raise ConanInvalidConfiguration("gcd eventloop is only supported on Linux with clang")
+
+        if self.options.eventloop == "asio" and tools.Version(self.version) < "0.8.0":
+            raise ConanInvalidConfiguration("asio eventloop is only supported on uSockets >= 0.8.0")
```
Comment (madebr) [inline recipes/usockets/all/conanfile.py:66]:
```suggestion
        if tools.Version(self.version) < "0.8.0" and self.options.eventloop not in ("libuv", "gcd"):
            raise ConanInvalidConfiguration(f"eventloop={self.options.eventloop} is not supported with {self.name}/{self.version}")
```

File: recipes/usockets/all/conanfile.py [hunk 4]
```diff
@@ -26,17 +28,77 @@ def config_options(self):
...
+
+        cppstd = self._cppstd
+        if cppstd == False:
```
Comment (madebr) [inline recipes/usockets/all/conanfile.py:69]:
```suggestion
        if not cppstd:
```

File: recipes/usockets/all/conanfile.py [hunk 5]
```diff
@@ -26,17 +28,77 @@ def config_options(self):
...
+            return
+        cppstd = str(cppstd)
```
Comment (madebr) [inline recipes/usockets/all/conanfile.py:71]: I don't think this is needed if you'ld return integers instead in `_cppstd` instead.

File: recipes/usockets/all/conanfile.py (hunk 1 above)
Comment (toge) [reply recipes/usockets/all/conanfile.py:17]: I agree with you.
I modify eventloop's default value to "syscall" which is used on Linux(epoll), MacOSX(kqueue).
On Windows, default value is overrided  by "libuv".

File: recipes/usockets/all/conanfile.py (hunk 2 above)
Comment (toge) [reply recipes/usockets/all/conanfile.py:100]: Yes, uSocket uses boost.asio as asio.
This option value is misleading.
I change it.

File: recipes/usockets/all/conanfile.py (hunk 4 above)
Comment (toge) [reply recipes/usockets/all/conanfile.py]: I agree with you.
It's my bad custom.

File: recipes/usockets/all/conanfile.py (hunk 5 above)
Comment (toge) [reply recipes/usockets/all/conanfile.py:71]: I agree with you.

File: recipes/usockets/all/conanfile.py (hunk 3 above)
Comment (toge) [reply recipes/usockets/all/conanfile.py:66]: I agree with you.
Your code is more generic than me.

I add "syscall" to "not in" condition.
```
        if tools.Version(self.version) < "0.8.0" and self.options.eventloop not in ("syscall", "libuv", "gcd"):
```

**Comment by toge** [comment]
Still license for Visual Studio has expired.
Is it my fault?

**Comment by madebr** [comment]
No worries, you didn't break anything :smile: 
It's a problem with the C3I servers.
I have also recipes that currently fail to build.

@jgsogo

**Comment by toge** [comment]
Thank you, @madebr .
I am wait for recovery.

**Comment by jgsogo** [comment]
>   **********************************************************************
>   ** Visual Studio 2017 Developer Command Prompt v15.9.19
>   ** Copyright (c) 2017 Microsoft Corporation
>   **********************************************************************
>   [vcvarsall.bat] Environment initialized for: 'x64'
>   
>   Microsoft Visual Studio 2017 Version 15.0.28307.1000.
>   Copyright (C) Microsoft Corp. All rights reserved.
>   
>   The license for Visual Studio has expired.
>   

😫  I'm on it! I'll trigger this PR later.

**Review by prince-chrismc — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Pull request merged**